File: app/src/main/java/com/weatherwise/api/WeatherRepository.java
package com.weatherwise.api;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.weatherwise.model.CurrentWeather;
import com.weatherwise.model.CurrentWeatherResponse;
import com.weatherwise.model.GeocodingResponse;
import com.weatherwise.model.OneCallResponse;
import com.weatherwise.service.WeatherParser;
import com.weatherwise.util.AppConstants;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class WeatherRepository {

    private static final String TAG = "WeatherRepository";

    private final WeatherApiService weatherService;
    private final WeatherApiService oneCallService;
    private final WeatherApiService geocodingService;

    // ── LiveData yang diobservasi ViewModel ────────────────────
    private final MutableLiveData<CurrentWeather> _currentWeather = new MutableLiveData<>();
    private final MutableLiveData<OneCallResponse> _oneCallData   = new MutableLiveData<>();
    private final MutableLiveData<Boolean>  _isLoading            = new MutableLiveData<>(false);
    private final MutableLiveData<String>   _errorMessage         = new MutableLiveData<>();
    private final MutableLiveData<String>   _currentCityName      = new MutableLiveData<>("");

    // Immutable untuk ViewModel
    public LiveData<CurrentWeather>  currentWeather = _currentWeather;
    public LiveData<OneCallResponse> oneCallData    = _oneCallData;
    public LiveData<Boolean>         isLoading      = _isLoading;
    public LiveData<String>          errorMessage   = _errorMessage;
    public LiveData<String>          currentCityName = _currentCityName;

    // Simpan koordinat terakhir untuk refresh
    private double lastLat = 0;
    private double lastLon = 0;
    private String units = AppConstants.UNITS;

    public WeatherRepository() {
        weatherService   = RetrofitClient.getWeatherService();
        oneCallService   = RetrofitClient.getOneCallService();
        geocodingService = RetrofitClient.getGeocodingService();
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Search berdasarkan nama kota
    // Alur: geocode → getCurrentWeather → getOneCall
    // ──────────────────────────────────────────────────────────
    public void fetchWeatherByCity(String cityName) {
        _isLoading.setValue(true);
        _errorMessage.setValue(null);

        geocodingService.geocodeCity(
                cityName,
                5,
                AppConstants.API_KEY
        ).enqueue(new Callback<List<GeocodingResponse>>() {

            @Override
            public void onResponse(Call<List<GeocodingResponse>> call,
                                   Response<List<GeocodingResponse>> response) {
                if (response.isSuccessful()
                        && response.body() != null
                        && !response.body().isEmpty()) {

                    GeocodingResponse geo = response.body().get(0);
                    lastLat = geo.lat;
                    lastLon = geo.lon;

                    // Simpan nama kota untuk ditampilkan di UI
                    _currentCityName.setValue(geo.getDisplayName());

                    // Lanjut fetch cuaca dengan koordinat hasil geocode
                    fetchWeatherByCoord(geo.lat, geo.lon);

                } else {
                    _isLoading.setValue(false);
                    _errorMessage.setValue(
                            "Kota \"" + cityName + "\" tidak ditemukan. " +
                                    "Coba periksa ejaan atau gunakan nama dalam Bahasa Inggris."
                    );
                }
            }

            @Override
            public void onFailure(Call<List<GeocodingResponse>> call, Throwable t) {
                _isLoading.setValue(false);
                _errorMessage.setValue(handleNetworkError(t));
                Log.e(TAG, "Geocoding gagal: " + t.getMessage(), t);
            }
        });
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Search berdasarkan koordinat GPS
    // ──────────────────────────────────────────────────────────
    public void fetchWeatherByCoord(double lat, double lon) {
        _isLoading.setValue(true);
        lastLat = lat;
        lastLon = lon;

        weatherService.getCurrentWeatherByCoord(
                lat, lon,
                AppConstants.API_KEY,
                units,
                AppConstants.LANG
        ).enqueue(new Callback<CurrentWeatherResponse>() {

            @Override
            public void onResponse(Call<CurrentWeatherResponse> call,
                                   Response<CurrentWeatherResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    // Parse response → model UI
                    CurrentWeather parsed =
                            WeatherParser.parseCurrentWeather(response.body());
                    _currentWeather.setValue(parsed);

                    // Lanjut fetch OneCall untuk hourly + daily + alerts
                    fetchOneCallData(lat, lon);

                } else {
                    _isLoading.setValue(false);
                    _errorMessage.setValue(parseHttpError(response.code()));
                }
            }

            @Override
            public void onFailure(Call<CurrentWeatherResponse> call, Throwable t) {
                _isLoading.setValue(false);
                _errorMessage.setValue(handleNetworkError(t));
                Log.e(TAG, "getCurrentWeather gagal: " + t.getMessage(), t);
            }
        });
    }

    // ──────────────────────────────────────────────────────────
    // PRIVATE: Fetch One Call (hourly + daily + alerts)
    // Dipanggil setelah getCurrentWeather berhasil
    // ──────────────────────────────────────────────────────────
    private void fetchOneCallData(double lat, double lon) {
        oneCallService.getOneCallData(
                lat, lon,
                AppConstants.API_KEY,
                units,
                AppConstants.LANG,
                "minutely"           // exclude minutely → hemat bandwidth
        ).enqueue(new Callback<OneCallResponse>() {

            @Override
            public void onResponse(Call<OneCallResponse> call,
                                   Response<OneCallResponse> response) {
                // Loading selesai di sini — ini call terakhir
                _isLoading.setValue(false);

                if (response.isSuccessful() && response.body() != null) {
                    _oneCallData.setValue(response.body());
                } else {
                    // OneCall gagal tidak fatal — cuaca utama sudah ada
                    // Cukup log, tidak perlu tampilkan error ke user
                    Log.w(TAG, "OneCall response gagal: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<OneCallResponse> call, Throwable t) {
                _isLoading.setValue(false);
                // OneCall gagal tidak fatal, cuaca utama tetap tampil
                Log.w(TAG, "OneCall request gagal: " + t.getMessage());
            }
        });
    }


    // ──────────────────────────────────────────────────────────
    // PUBLIC: Set unit suhu ("metric" / "imperial")
    // ──────────────────────────────────────────────────────────
    public void setUnits(String newUnits) {
        if (newUnits == null || newUnits.trim().isEmpty()) return;
        units = newUnits.trim();
    }

    public String getUnits() {
        return units;
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Refresh — panggil ulang dengan koordinat terakhir
    // ──────────────────────────────────────────────────────────
    public void refresh() {
        if (lastLat != 0 || lastLon != 0) {
            fetchWeatherByCoord(lastLat, lastLon);
        } else {
            _errorMessage.setValue("Belum ada data untuk di-refresh.");
        }
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Hapus pesan error
    // ──────────────────────────────────────────────────────────
    public void clearError() {
        _errorMessage.setValue(null);
    }

    // ──────────────────────────────────────────────────────────
    // PRIVATE: Helper — terjemahkan HTTP error code → pesan user
    // ──────────────────────────────────────────────────────────
    private String parseHttpError(int code) {
        switch (code) {
            case 401: return "API Key tidak valid. Periksa konfigurasi AppConstants.";
            case 404: return "Data tidak ditemukan untuk lokasi ini.";
            case 429: return "Terlalu banyak permintaan. Coba beberapa saat lagi.";
            case 500:
            case 502:
            case 503: return "Server OpenWeatherMap sedang bermasalah. Coba lagi nanti.";
            default:  return "Terjadi kesalahan (kode " + code + "). Coba lagi.";
        }
    }

    // ──────────────────────────────────────────────────────────
    // PRIVATE: Helper — terjemahkan network error → pesan user
    // ──────────────────────────────────────────────────────────
    private String handleNetworkError(Throwable t) {
        if (t instanceof java.net.UnknownHostException) {
            return "Tidak ada koneksi internet. Periksa jaringanmu.";
        } else if (t instanceof java.net.SocketTimeoutException) {
            return "Koneksi timeout. Jaringan terlalu lambat, coba lagi.";
        } else if (t instanceof java.io.IOException) {
            return "Gagal terhubung ke server: " + t.getMessage();
        }
        return "Terjadi kesalahan tidak terduga: " + t.getMessage();
    }
}

File: app/src/main/java/com/weatherwise/api/WeatherRepository.kt
package com.weatherwise.api

import android.content.Context
import android.util.Log
import com.weatherwise.data.CachedWeatherEntity
import com.weatherwise.data.WeatherDatabase
import com.weatherwise.model.CurrentWeather
import com.weatherwise.model.CurrentWeatherResponse
import com.weatherwise.model.ForecastDaily
import com.weatherwise.model.ForecastHourly
import com.weatherwise.model.GeocodingResponse
import com.weatherwise.service.WeatherParser
import com.weatherwise.util.AppConstants
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import retrofit2.HttpException
import java.io.IOException

class WeatherRepository(private val context: Context) {
    private val weatherService = RetrofitClient.getWeatherServiceKotlin()
    private val weatherDao = WeatherDatabase.getDatabase(context).weatherDao()

    private val _currentWeather = MutableStateFlow<CurrentWeather?>(null)
    val currentWeather: StateFlow<CurrentWeather?> = _currentWeather.asStateFlow()

    private val _hourlyForecast = MutableStateFlow<List<ForecastHourly>>(emptyList())
    val hourlyForecast: StateFlow<List<ForecastHourly>> = _hourlyForecast.asStateFlow()

    private val _dailyForecast = MutableStateFlow<List<ForecastDaily>>(emptyList())
    val dailyForecast: StateFlow<List<ForecastDaily>> = _dailyForecast.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _currentCityName = MutableStateFlow("")
    val currentCityName: StateFlow<String> = _currentCityName.asStateFlow()

    private var lastLat = 0.0
    private var lastLon = 0.0
    private var units = AppConstants.UNITS

    suspend fun fetchWeatherByCity(cityName: String) {
        _isLoading.value = true
        _errorMessage.value = null

        try {
            val geoResponse = weatherService.geocodeCity(cityName, 5, AppConstants.API_KEY)
            if (geoResponse.isSuccessful && geoResponse.body()?.isNotEmpty() == true) {
                val geo = geoResponse.body()!![0]
                lastLat = geo.lat
                lastLon = geo.lon
                _currentCityName.value = geo.getDisplayName()
                fetchWeatherByCoord(geo.lat, geo.lon)
            } else {
                _isLoading.value = false
                _errorMessage.value = "Kota \"$cityName\" tidak ditemukan."
            }
        } catch (e: Exception) {
            _isLoading.value = false
            _errorMessage.value = handleNetworkError(e)
            Log.e("WeatherRepository", "Geocoding failed", e)
        }
    }

    suspend fun fetchWeatherByCoord(lat: Double, lon: Double) {
        _isLoading.value = true
        lastLat = lat
        lastLon = lon

        try {
            // 1. Fetch Current Weather
            val currentResponse = weatherService.getCurrentWeatherByCoord(lat, lon, AppConstants.API_KEY, units, AppConstants.LANG)
            if (currentResponse.isSuccessful && currentResponse.body() != null) {
                val parsed = WeatherParser.parseCurrentWeather(currentResponse.body()!!)
                _currentWeather.value = parsed
                
                if (!parsed.cityName.isNullOrEmpty()) {
                    _currentCityName.value = parsed.cityName!!
                }

                // Cache to Room
                weatherDao.insertCachedWeather(CachedWeatherEntity(
                    cityName = parsed.cityName ?: "Unknown",
                    lat = lat,
                    lon = lon,
                    temperature = parsed.temperature,
                    condition = parsed.condition ?: "",
                    conditionIcon = parsed.conditionIcon ?: "",
                    humidity = parsed.humidity,
                    windSpeed = parsed.windSpeed,
                    timestamp = System.currentTimeMillis()
                ))

                // 2. Fetch Forecast (replaces OneCall for free tier compatibility)
                val forecastResponse = weatherService.getForecast(lat, lon, AppConstants.API_KEY, units, AppConstants.LANG)
                if (forecastResponse.isSuccessful && forecastResponse.body() != null) {
                    parseForecastData(forecastResponse.body()!!)
                } else {
                    Log.w("WeatherRepository", "Forecast failed: ${forecastResponse.code()}")
                }
            } else {
                _errorMessage.value = parseHttpError(currentResponse.code())
            }
        } catch (e: Exception) {
            _errorMessage.value = handleNetworkError(e)
            Log.e("WeatherRepository", "Fetch weather failed", e)
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun loadCachedWeather() {
        val cached = weatherDao.getCachedWeather()
        if (cached != null) {
            // Reconstruct a basic CurrentWeather object from cache
            val cachedWeather = CurrentWeather().apply {
                this.cityName = cached.cityName
                this.temperature = cached.temperature
                this.condition = cached.condition
                this.conditionIcon = cached.conditionIcon
                this.humidity = cached.humidity
                this.windSpeed = cached.windSpeed
            }
            _currentWeather.value = cachedWeather
            _currentCityName.value = cached.cityName
            lastLat = cached.lat
            lastLon = cached.lon
            
            // Trigger a background refresh
            fetchWeatherByCoord(cached.lat, cached.lon)
        }
    }

    private fun parseForecastData(response: com.weatherwise.model.ForecastResponse) {
        // Map to hourly (next 24 hours = 8 items of 3 hours)
        val hourly = response.list.take(8).map { item ->
            ForecastHourly(
                dt = item.dt,
                temp = item.main.temp,
                condition = item.weather[0].main,
                conditionIcon = item.weather[0].icon,
                humidity = item.main.humidity,
                windSpeed = item.wind.speed
            )
        }
        _hourlyForecast.value = hourly

        // Map to daily (group by date, get min/max)
        val dailyMap = mutableMapOf<String, ForecastDaily>()
        for (item in response.list) {
            val dateKey = item.dt_txt.substring(0, 10) // "YYYY-MM-DD"
            val temp = item.main.temp
            val condition = item.weather[0].main
            val icon = item.weather[0].icon
            
            if (!dailyMap.containsKey(dateKey)) {
                dailyMap[dateKey] = ForecastDaily(
                    dt = item.dt,
                    dateStr = dateKey,
                    tempMax = temp,
                    tempMin = temp,
                    condition = condition,
                    conditionIcon = icon,
                    humidity = item.main.humidity
                )
            } else {
                val existing = dailyMap[dateKey]!!
                dailyMap[dateKey] = existing.copy(
                    tempMax = maxOf(existing.tempMax, temp),
                    tempMin = minOf(existing.tempMin, temp)
                )
            }
        }
        _dailyForecast.value = dailyMap.values.toList().take(7)
    }

    private fun parseHttpError(code: Int): String = when (code) {
        401 -> "API Key tidak valid."
        404 -> "Data tidak ditemukan."
        429 -> "Terlalu banyak permintaan."
        else -> "Terjadi kesalahan (kode $code)."
    }

    private fun handleNetworkError(e: Exception): String = when (e) {
        is IOException -> "Tidak ada koneksi internet."
        is HttpException -> "Gagal terhubung ke server."
        else -> "Terjadi kesalahan: ${e.message}"
    }

    fun setUnits(newUnits: String) {
        units = newUnits
    }

    suspend fun refresh() {
        if (lastLat != 0.0 || lastLon != 0.0) {
            fetchWeatherByCoord(lastLat, lastLon)
        } else {
            _errorMessage.value = "Belum ada data untuk di-refresh."
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }

    suspend fun fetchGeocoding(query: String): List<GeocodingResponse> {
        return try {
            val response = weatherService.geocodeCity(query, 5, AppConstants.API_KEY)
            if (response.isSuccessful && response.body() != null) {
                response.body()!!
            } else {
                emptyList()
            }
        } catch (e: Exception) {
            Log.e("WeatherRepository", "Geocoding search failed", e)
            emptyList()
        }
    }
}
